rel/marathon.cpp

```diff
@@ -0,0 +1,120 @@
+
+    ball.vel = entered_goal_vel_rt_start;  // Apply it to the ball! (finally)
+    // Now that it's used, get rid of the stored vector
+
+    // MATRIX NOTES FROM COMPLEX (THE GOAT) BELOW vvv
+    // Vec pos_rt_itemgroup = {5, 5, 0};
+    // Vec pos_rt_world = {};
+    // mtxa = [ig_to_world]
+    // mkb::mtxa_from_mtx(&mkb::itemgroups[0].transform);
+    // pos_rt_world = mtxa * pos_rt_itemgroup
+    // mkb::mtxa_tf_point(&pos_rt_itemgroup, &pos_rt_world);
+} // apply_saved_vel
+
+// [Nambo] Transforms the current ball velocity relative to the goal, and stores it
+static void store_saved_vel() {
+    mkb::Ball& ball = mkb::balls[mkb::curr_player_idx];  // I think this makes ball.vel work
+
+    // Gives names to the item group index for the goal's IG, and the goal's index inside that IG
+    // This keeps the [] stuff inside the TF code short!
+    s16 goal_idx = mkb::mode_info.entered_goal_idx;
+    s16 goalIG_idx = mkb::mode_info.entered_goal_itemgroup_idx;
+
+    // MATRIX NOTES FROM COMPLEX (THE GOAT) BELOW vvv
+    // Build [world_from_goal]
+    // Build [world_from_itemgroup] * [itemgroup_from_goal]
+    // Build [world_from_itemgroup] * [translation(goal)] * [zrotation(goal)] * [yrotation(goal)] *
+    // [xrotation(goal)]
+
+    // Set up the TF matrix for Goal ---> World (Inverse of World ---> Goal)
+    mkb::mtxa_from_mtx(&mkb::itemgroups[goalIG_idx].transform);  // Build [world_from_itemgroup]
+    mkb::mtxa_translate(&mkb::stagedef->coli_header_list[goalIG_idx].goal_list[goal_idx].position);
+    mkb::mtxa_rotate_z(mkb::stagedef->coli_header_list[goalIG_idx].goal_list[goal_idx].rotation.z);
+    mkb::mtxa_rotate_y(mkb::stagedef->coli_header_list[goalIG_idx].goal_list[goal_idx].rotation.y);
+    mkb::mtxa_rotate_x(mkb::stagedef->coli_header_list[goalIG_idx].goal_list[goal_idx].rotation.x);
+    // Applies this TF ^^^
+    Vec entered_goal_vel_rt_world = {ball.vel.x, ball.vel.y, ball.vel.z};
+    Vec entered_goal_vel_rt_goal = {};
+    mkb::mtxa_rigid_inv_tf_vec(&entered_goal_vel_rt_world, &entered_goal_vel_rt_goal);
+    // We have now stored the desired veloocity vector as entered_goal_vel_rt_goal
+    s_saved_vel = entered_goal_vel_rt_goal;
+} // store_saved_vel
+
+void tick() {
+    if(pref::get_marathon()) {
+        if(s_state == MarathonState::WaitForGoal) {
+            if(mkb::sub_mode == mkb::SMD_GAME_GOAL_INIT || mkb::sub_mode == mkb::SMD_GAME_GOAL_MAIN) {
+                s_state = MarathonState::StoringVel;
+            }
+        } // State 1: WaitForGoal
+        else if(s_state == MarathonState::StoringVel) {
+            store_saved_vel();
+            s_state = MarathonState::WaitForFirstApply;
+        } // State 2: StoringVel
+        else if(s_state == MarathonState::WaitForFirstApply) {
+            if(mkb::mode_info.stage_time_frames_remaining == mkb::mode_info.stage_time_limit - 1) {
+                apply_saved_vel();
+                s_state = MarathonState::WaitForApplyOrGoal;
+            }
```

Comment:
Since the exact same check-goal-and-save functionality is used in two different states, I'd factor it out to a separate function. If you wanted, it might make sense to put the implementation of `apply_saved_vel()` directly in this func as well.